<h1 align=center><font size = 5>Final Submission</font></h1>

* **course:** Introduction to Deep Learning & Neural Networks with Keras
* **submission:** regression model in Keras for concrete compressive strength
* **my name:** Zong Hui(宗辉)

importing libraries:

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers.convolutional import Conv2D # to add convolutional layers
from keras.layers.convolutional import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers

Using TensorFlow backend.


loading data:

In [2]:
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

***

## A. Build a baseline model 

Use the Keras library to build a neural network with the following:

  - One hidden layer of 10 nodes, and a ReLU activation function

  - Use the adam optimizer and the mean squared error as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

Use the Keras library to build a neural network:

In [4]:
# define baseline model
def baseline_model():
    # One hidden layer of 10 nodes, and a ReLU activation function
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(8,)))
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

1.Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

In [5]:
concrete_data_columns = concrete_data.columns
X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
y = concrete_data["Strength"]

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [7]:
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (721, 8)
Shape of X_test: (309, 8)
Shape of y_train: (721,)
Shape of y_test: (309,)


2.Train the model on the training data using 50 epochs.

In [8]:
# build the model
model = baseline_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)

3.Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [9]:
predicted_concrete_strength = y_test
actual_concrete_strength = model.predict(X_test)
print(mean_squared_error(actual_concrete_strength, predicted_concrete_strength))

1852.7627825364384


4.Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [10]:
mean_squared_errors = []

for i in range(50):
    # 1.Randomly split the data into a training and test sets by holding 30% of the data for testing.
    concrete_data_columns = concrete_data.columns
    X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
    y = concrete_data["Strength"]
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)
    
    # 2.Train the model on the training data using 50 epochs.
    model = baseline_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # 3.Evaluate the model on the test data and compute the mean squared error
    predicted_concrete_strength = model.predict(X_test)
    actual_concrete_strength = y_test
    mean_squared_error_score = mean_squared_error(actual_concrete_strength, predicted_concrete_strength)
    mean_squared_errors.append(mean_squared_error_score)

5.Report the mean and the standard deviation of the mean squared errors.

In [11]:
print("mean of the mean squared errors: ", np.mean(mean_squared_errors))
print("standard deviation of the mean squared errors: ", np.std(mean_squared_errors))

mean of the mean squared errors:  371.3202938531652
standard deviation of the mean squared errors:  384.6245572864479


***

## B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [12]:
# define baseline model
def baseline_model():
    # One hidden layer of 10 nodes, and a ReLU activation function
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(8,)))
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [13]:
mean_squared_errors = []

for i in range(50):
    # 1.Randomly split the data into a training and test sets by holding 30% of the data for testing.
    concrete_data_columns = concrete_data.columns
    X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
    y = concrete_data["Strength"]
    X_norm = (X - X.mean()) / X.std()
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.30)
    
    # 2.Train the model on the training data using 50 epochs.
    model = baseline_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # 3.Evaluate the model on the test data and compute the mean squared error
    predicted_concrete_strength = model.predict(X_test)
    actual_concrete_strength = y_test
    mean_squared_error_score = mean_squared_error(actual_concrete_strength, predicted_concrete_strength)
    mean_squared_errors.append(mean_squared_error_score)

In [14]:
print("mean of the mean squared errors: ", np.mean(mean_squared_errors))
print("standard deviation of the mean squared errors: ", np.std(mean_squared_errors))

mean of the mean squared errors:  385.364690535592
standard deviation of the mean squared errors:  84.45719199401023


### How does the mean of the mean squared errors compare to that from Step A?

371.3202938531652 changed to 385.364690535592

***

## C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

In [15]:
# define baseline model
def baseline_model():
    # One hidden layer of 10 nodes, and a ReLU activation function
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(8,)))
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [16]:
mean_squared_errors = []

for i in range(50):
    # 1.Randomly split the data into a training and test sets by holding 30% of the data for testing.
    concrete_data_columns = concrete_data.columns
    X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
    y = concrete_data["Strength"]
    X_norm = (X - X.mean()) / X.std()
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.30)
    
    # 2.Train the model on the training data using 50 epochs.
    model = baseline_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=0)
    
    # 3.Evaluate the model on the test data and compute the mean squared error
    predicted_concrete_strength = model.predict(X_test)
    actual_concrete_strength = y_test
    mean_squared_error_score = mean_squared_error(actual_concrete_strength, predicted_concrete_strength)
    mean_squared_errors.append(mean_squared_error_score)

In [17]:
print("mean of the mean squared errors: ", np.mean(mean_squared_errors))
print("standard deviation of the mean squared errors: ", np.std(mean_squared_errors))

mean of the mean squared errors:  158.99688714527093
standard deviation of the mean squared errors:  14.66178968505943


### How does the mean of the mean squared errors compare to that from Step B?

385.364690535592 changed to 158.99688714527093

***

## D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

In [18]:
# define baseline model
def baseline_model():
    # One hidden layer of 10 nodes, and a ReLU activation function
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(8,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # Use the adam optimizer and the mean squared error as the loss function.
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [19]:
mean_squared_errors = []

for i in range(50):
    # 1.Randomly split the data into a training and test sets by holding 30% of the data for testing.
    concrete_data_columns = concrete_data.columns
    X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
    y = concrete_data["Strength"]
    X_norm = (X - X.mean()) / X.std()
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.30)
    
    # 2.Train the model on the training data using 50 epochs.
    model = baseline_model()
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, verbose=0)
    
    # 3.Evaluate the model on the test data and compute the mean squared error
    predicted_concrete_strength = model.predict(X_test)
    actual_concrete_strength = y_test
    mean_squared_error_score = mean_squared_error(actual_concrete_strength, predicted_concrete_strength)
    mean_squared_errors.append(mean_squared_error_score)

In [20]:
print("mean of the mean squared errors: ", np.mean(mean_squared_errors))
print("standard deviation of the mean squared errors: ", np.std(mean_squared_errors))

mean of the mean squared errors:  128.26050065733313
standard deviation of the mean squared errors:  15.495405305120572


### How does the mean of the mean squared errors compare to that from Step B?

385.364690535592 changed to 128.26050065733313

***

### Thanks
thanks IBM for the **Coursera** called *Introduction to Deep Learning & Neural Networks with Keras*! I have learnt a lot!